### PR TITLE
Update QueryTranslationTest to use 5.1 topologies

### PR DESCRIPTION
**Pull request by vcrfxia**
### Description 

**Update:**

In light of the discussion below, this PR no longer updates QueryTranslationTest to test both with and without streams optimizations, since the streams optimization setting is not intended to be configurable by users. Instead, this PR only updates QueryTranslationTest to use newly generated 5.1 topologies.

**Original:**

This PR updates `TopologyFileGenerator` to create two `TestCase`s (for `QueryTranslationTest`) from each `JsonTestCase`, one with streams optimizations enabled and one with optimizations disabled.

This PR also adds new expected topology files (for KSQL 5.1), generated from the `5.1.x` branch, which are now checked against by `QueryTranslationTest`. Because 5.1 does not include streams optimizations, `5.1.x` could not be used to generate expected topology files for tests where streams optimizations are enabled. Instead, expected topology files for tests with streams optimizations enabled were generated from the current `master` branch. (Perhaps these expected topologies should be moved out of the `expected_topology/5_1` folder, to avoid confusion?)

Note that all files in `expected_topolgoy/5_1` were auto-generated by `TopologyFileGenerator` on the branch `5.1.x`. The `ksql.streams.topology.optimization` config was added to the generated files in order to enable or disable streams optimizations as appropriate, and the `ksql.named.internal.topics` config was similarly added, as setting `ksql.streams.topology.optimization=all` requires setting `ksql.named.internal.topics=on`.

Another necessary tweak was to explicitly specify `ksql.windowed.session.key.legacy=false` in two of the query translation tests, since these tests contain session windows with equal start and end times, which is incompatible with `ksql.windowed.session.key.legacy=true` (i.e., running the test results in a `NegativeArraySizeException`). There are actually more than two query translation tests that involve session windows with equal start and end times, but only the the tests that existed in 5.1 need this additional config explicitly set, because of how `QueryTranslationTest` builds configs. Specifically:

- If a topology file for the test exists, then configs are read from the top of the topology file. Topology files generated from a 5.1 branch do not have the `ksql.windowed.session.key.legacy` config set, even though it is a compatibility breaking config, since it was only introduced after 5.1. As a consequence, the inferred value for the `ksql.windowed.session.key.legacy` is the legacy value `true`. This behavior explains why we have to explicitly add `ksql.windowed.session.key.legacy=false` to the two session window tests mentioned above.

- If a topology file for the test does not exist, then the current default KsqlConfig is used. This config uses the current value of compatibility breaking configs, which is `false` for the `ksql.windowed.session.key.legacy` config. This explains why we do not have to add `ksql.windowed.session.key.legacy=false` to the tests that have session windows with equal start and end times that were not present in 5.1 (since they do not have generated topology files). However, perhaps we should explicitly add `ksql.windowed.session.key.legacy=false` to these tests for completeness anyway.

### Testing done 

Query translation tests continue to pass.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by apurvam** [comment]
Thanks for the great description @vcrfxia ! It makes sense.

One clarification which I think would be helpful is to be detailed about the precise steps for generating the topology files and creating this pr. Was it:

1. `git checkout 5.1.x`
2. Run `TopologyFileGenerator`
3. `git checkout master`
4. `git add generated_directory`.
5. run tests.
6. Add `ksql.windowed.session.key.legacy=false` to the two session window tests carried over from 5.1

?

I think that is all that you should have had to do. If you had to deviate from that, it would be good to understand why. Specifically, I still don't know why there should have been any special steps for handling `ksql.streams.topology.optimization`.

cc @bbejeck

**Comment by vcrfxia** [comment]
@apurvam To generate topology files from `5.1.x`, the steps are:

1. Checkout `5.1.x`.
2. Add the line `ksqlEngine.close();` after [line 602](https://github.com/confluentinc/ksql/blob/7a9a04ff69f5c5a84f75b636c094e82e3ca22fdc/ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java#L602) in `EndToEndEngineTestUtil.java`, or else running `TopologyFileGenerator.java` fails with the error `java.lang.IllegalArgumentException: A metric named 'MetricName [name=_confluent-ksql-some.ksql.service.idksql-engine-query-stats-CREATED-queries, group=ksql-engine-query-stats, description=Count of queries in CREATED state., tags={}]' already exists, can't register another one.`
3. At this point you can run `TopologyFileGenerator.java` just fine, but if you take those expected topologies and run `QueryTranslationTest` against them on `master`, more than 100 tests fail with the error `java.lang.RuntimeException: Internal topic naming must be enabled if streams optimizations enabled`. This is because the default configs used by `TopologyFileGenerator` and `QueryTranslationTest` include `ksql.streams.topology.optimization=all` but does not include `ksql.named.internal.topics` (see [here](https://github.com/confluentinc/ksql/blob/7a9a04ff69f5c5a84f75b636c094e82e3ca22fdc/ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java#L791)), since the latter does not exist on 5.1. However, on master we throw an error if someone tries creating a KsqlConfig with `ksql.streams.topology.optimization=all` and `ksql.named.internal.topics=off`, the latter of which is inferred from configs that don't have `ksql.named.internal.topics` since that's the legacy setting for the config. So one of two things needs to be done:

- Turn off streams optimizations by removing [line 791](https://github.com/confluentinc/ksql/blob/7a9a04ff69f5c5a84f75b636c094e82e3ca22fdc/ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java#L791) of `EndToEndEngineTestUtil` where it's currently being added.
- Set `ksql.named.internal.topics=on` by adding the relevant config into the block around [line 791](https://github.com/confluentinc/ksql/blob/7a9a04ff69f5c5a84f75b636c094e82e3ca22fdc/ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java#L791). If you go with this option, you also have to modify `KsqlConfigResolver`/`KsqlConfig` to accept the config, otherwise it's ignored. (This happens because `KsqlConfigResolver` is set up to ignore unrecognized configs, and `ksql.named.internal.topics` does not exist in 5.1.)

4. Now run `TopologyFileGenerator.java`.
5. Checkout `master`.
6. Update [line 111](https://github.com/confluentinc/ksql/blob/a1f3e03c87dd7a40fb58b7d63a6fead501e82814/ksql-engine/src/test/java/io/confluent/ksql/QueryTranslationTest.java#L111) of `QueryTranslationTest` to point to the new folder of expected topologies.
7. Run `QueryTranslationTest`, and you'll see two test failures, the two session window tests discussed in the PR description above. Apply the fix for those, and you're done.

The good news is we don't have to include the code changes for fixing these various hiccups in this PR since they're all already fixed on `master` (with the exception of the fix for the two session window tests); the fixes just weren't on `5.1.x`. In particular, the hiccup in step 2 was fixed by https://github.com/confluentinc/ksql/pull/2243, and the hiccup in step 3 was fixed by https://github.com/confluentinc/ksql/pull/2077. The usual steps for updating the expected topologies should be more along the lines of:
```
git checkout <old_branch>
run TopologyFileGenerator
git checkout master
update QueryTranslationTest to point to the new directory of expected topologies
```
and that's it.

The additional code changes in this PR are simply to support testing with _both_ streams optimizations enabled and disabled. If not for that, this PR wouldn't have code changes at all (besides pointing `QueryTranslationTest` to the new expected topologies.

**Comment by apurvam** [comment]
Thanks for that amazingly detailed description, @vcrfxia . It is indeed very clear about what you did and why you had to do it.

@confluentinc/ksql, @confluentinc/streams , this brings up good questions around how we plan to manage backward compatible upgrades when we add optimizations in the future. 

1. How we envision using the `streams.topology.optimization` config. Right now it takes config values `all` and `off`. I guess there is a plan to version the optimizations? 
2. It seems to me that when we do add optimizations, we will probably add additional configs control related behavior, like named internal topics in the present case. This means that '3' above is probably going to be a constant operation we need to perform when we add new versions. Is that a reasonable expectation?

Other than that, this change set looks good to me. Thanks again for taking the time to write that explanation up, @vcrfxia , I think it is very valuable.

**Comment by big-andy-coates** [comment]
I'm going to bow out of this PR as its not an area i know enough to comment on.

File: ksql-engine/src/test/resources/query-validation-tests/session-windows.json [hunk 1]
```diff
@@ -22,7 +22,10 @@
         {"topic": "S2", "key": 1, "value": "1,100", "timestamp": 10000, "window": {"start": 10000, "end": 10000, "type": "session"}},
         {"topic": "S2", "key": 1, "value": null, "timestamp": 40000, "window": {"start": 10000, "end": 10000, "type": "session"}},
         {"topic": "S2", "key": 1, "value": "1,200", "timestamp": 40000, "window": {"start": 10000, "end": 40000, "type": "session"}}
-      ]
+      ],
+      "properties": {
+        "ksql.windowed.session.key.legacy": "false"
```
Comment (big-andy-coates) [inline ksql-engine/src/test/resources/query-validation-tests/session-windows.json]: Yep, this is what I was talking about :D

**Comment by apurvam** [comment]
> Currently, streams uses a one-up numbering scheme when creating a topology. So a shift in the number of processing nodes causes the internal numbering to change, but from a user standpoint as long as the changelog and reparition topics don't change, the new toplogy should be compatible.

So if a future optimization collapses topology nodes, the topology will be different, even though there is no practical difference? So the current upgrade test will fail in that case, or does the comparator account for this?

**Comment by bbejeck** [comment]
@apurvam 

Sorry I think I've gone in the wrong direction and I'm not very clear in addressing your concerns, let me take a step back.

To answer your question on topology comparison, it's just a straight string compare of the generated topology against the expected topology.  Any changes from Streams should not change the internal numbering on the final topology; otherwise, it's a regression on Streams that needs to be addressed. 

To go back to your original question point 3 from @vcrfxia's description above, yes generating expected topology files
will be an ongoing issue.

The critical question is how often to generate topology files? We don't want each change made to require a rebuild of topology files, but we 
want them frequent enough that they keep up with changes being made (either KSQL or Streams).

My initial thought and implementation of the file generation assumed that we would initially generate some expected topology and at the very end of a development cycle for a given and copy those files to a folder representing
the version just released. 

Then during the development of the next release, depending on features added by KSQL, new expected topology files may need to get
generated.  This ideally will be an automated process, but again at what interval?  We've discussed this before and didn't settle on an
on anything, so it's a manual process at the moment.

So until we identify an automated solution here's what my thoughts are.

1.  The current suite of unit tests will always run the expected topology files against the current target release 5_2_0 in this case.  When adding new features that either modify or create a new topology, we'll need to regenerate new topology files. For right now this is a manual process, but hopefully, with the `TopologyFileGenerator` it's just push-button process.

2. At the end of a release, those generated topology files are copied from the folder for the current version into a new folder for the new upcoming release.  Then the 
topology files from the previous release are considered "frozen," and any additional topology file generation occurs in the new folder.

For example we should have a directory named `/src/test/resources/expected_topology/5_2_0` for the current release.  Unit test runs will always run the expected topology files for the upcoming release, so `QueryTranslationTest` will always compare what's being developed in master against the 5_2_0 folder.

Naturally, we also want to compare against older topology versions to ensure backward compatibility.  Right now the unit tests against a single release of expected topology files are pretty fast.  But over time unit test running against all previous releases will be too time-consuming for daily development activities.

One I idea is that we would have a script that will run all tests for each release under the `expected_topology` directory.  
Something along these lines:

```
for each dir in expected_topology
  do
     mvn test -Dtopology.version=$dir -Dtest=QueryTranslationTest -pl ksql-engine
 done
 ```
While having a script of this nature solves the unit test runs issue, it introduces another problem in the fact that running this script is again a manual process.  So maybe we could update the night KSQL build to run this script.

Does this make sense?

\cc @vcrfxia @rodesai

**Comment by rodesai** [comment]
We don't need to test with optimizations both enabled and disabled. Each version of KSQL should have a single possible value for the setting (though atm this is not enforced by the code, which we should add. Also, note that this is the case for other compatibility-breaking configs as well) - off for 5.1, on for 5.2 onwards, and version number if/when that gets added to streams.

As far as the cadence at which we generate topology files - I agree with @bbejeck here. We can generate as we release for now, and test against every previous release. This will need to be rethought as we release more frequently. One idea here is to add a test that verifies that we haven't introduced any compatibility-breaking changes, and then either fix the compatibility-breaking change or generate new files at that time. The way this test would work would be to produce topologies using the current default configs, and compare to the topologies generated using the latest topology save-point. If the topologies differ, the test fails.

**Comment by vcrfxia** [comment]
I see, thanks for clarifying @rodesai . I didn't realize the plan was to only allow KSQL to use a single value for the streams optimization setting at a time. In light of that, I agree it doesn't make sense to have query translation tests with and without optimizations at the same time. Pushed a change to remove the version with optimizations on.

What do you think? @apurvam

**Pull request merged**